File: module-07-transport-layer/tcp_echo_client.js
/**
 * Module 07 — JS/Node port of tcp_echo_client.py.
 */
const net = require("net");

const HOST = "127.0.0.1";
const PORT = 9017;

const socket = net.createConnection(PORT, HOST, () => {
  console.log(`Connected to ${HOST}:${PORT}`);
  sendNext(1);
});

function sendNext(i) {
  if (i > 5) {
    console.log("Closing connection (watch for TIME_WAIT on the server side now).");
    socket.end();
    return;
  }
  const message = `message #${i}`;
  socket.once("data", (data) => {
    const reply = data.toString();
    const status = reply === message ? "OK, matches" : "MISMATCH!";
    console.log(`  sent ${JSON.stringify(message)} -> got back ${JSON.stringify(reply)}  (${status})`);
    sendNext(i + 1);
  });
  socket.write(message);
}

File: module-07-transport-layer/tcp_echo_client.py
"""
Module 07 -- TCP echo client. Sends 5 numbered messages and confirms every
single one comes back, in order, unmodified.
"""
import socket

HOST, PORT = "127.0.0.1", 9007


def main():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.connect((HOST, PORT))
        print(f"Connected to {HOST}:{PORT}")

        for i in range(1, 6):
            message = f"message #{i}".encode()
            sock.sendall(message)
            reply = sock.recv(1024)
            status = "OK, matches" if reply == message else "MISMATCH!"
            print(f"  sent {message!r} -> got back {reply!r}  ({status})")

        print("Closing connection (watch for TIME_WAIT on the server side now).")


if __name__ == "__main__":
    main()
